keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -56,30 +67,156 @@ class NotificationEventListener(
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackUnpinnedNotification(
+            channel = it.address,
+            latestArtifact = latestArtifact,
+            pinnedArtifact = pinnedArtifact,
+            application = config.application,
+            time = clock.instant(),
+            user = user,
+            targetEnvironment = targetEnvironment
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(MarkAsBadNotification::class)
+  fun onMarkAsBadNotification(notification: MarkAsBadNotification) {
+    with(notification) {
+      val deliveryArtifact = config.artifacts.find {
+        it.reference == veto.reference
+      } ?: return
+
+      val vetoedArtifact = repository.getArtifactVersion(deliveryArtifact, veto.version, null)
+      if (vetoedArtifact == null) {
+        log.warn("vetoedArtifact is null for application ${config.application}. Can't send MarkAsBadNotification notification")
+        return
+      }
+
+      val handler = handlers.find {
+        it.type == notification.type
+      } as? SlackNotificationHandler<SlackMarkAsBadNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackMarkAsBadNotification(
+            vetoedArtifact = vetoedArtifact,
+            user = user,
+            targetEnvironment = veto.targetEnvironment,
+            channel = it.address,
+            time = clock.instant(),
+            application = config.name,
+            comment = veto.comment
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(ApplicationActuationPaused::class)
+  fun onApplicationActuationPaused(notification: ApplicationActuationPaused) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      val handler = handlers.find {
+        it.type == Type.PAUSED_APPLICATION
+      } as? SlackNotificationHandler<SlackPausedNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackPausedNotification(
+            user = triggeredBy,
+            channel = it.address,
+            time = clock.instant(),
+            application = application
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(ApplicationActuationResumed::class)
+  fun onApplicationActuationResumed(notification: ApplicationActuationResumed) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      val handler = handlers.find {
+        it.type == Type.PAUSED_APPLICATION
+      } as? SlackNotificationHandler<SlackResumedNotification> ?: return
+
+      config.environments.forEach { environment ->
```

Comment:
Same comment/suggestion on looping over environments.

Reply:
we should loop on the entire environments here, as it applies for all of them (correct me if I'm wrong)

Reply:
Oh, right.

Reply:
I prefer to keep it as is for now.. I think it's more readable.

Reply:
oops, you're right @luispollo . will change it.

Reply:
done